src/poll_manager.ts

```diff
@@ -0,0 +1,149 @@
+import type { StreamChat } from './client';
+import type {
+  CreatePollData,
+  DefaultGenerics,
+  ExtendableGenerics,
+  PollResponse,
+  PollSort,
+  QueryPollsFilters,
+  QueryPollsOptions,
+} from './types';
+import { Poll } from './poll';
+import { FormatMessageResponse } from './types';
+import { formatMessage } from './utils';
+
+export class PollManager<SCG extends ExtendableGenerics = DefaultGenerics> {
+  private client: StreamChat<SCG>;
+  private pollCache = new Map<string, Poll<SCG>>();
+  private unsubscribeFunctions: Set<() => void> = new Set();
+
+  constructor({ client }: { client: StreamChat<SCG> }) {
+    this.client = client;
+  }
+
+  get data(): Map<string, Poll<SCG>> {
+    return this.pollCache;
+  }
+
+  public fromState = (id: string) => {
+    return this.pollCache.get(id);
+  };
+
+  public registerSubscriptions = () => {
+    if (this.unsubscribeFunctions.size) {
+      // Already listening for events and changes
+      return;
+    }
+
+    this.unsubscribeFunctions.add(this.subscribeMessageNew());
+    this.unsubscribeFunctions.add(this.subscribePollUpdated());
+    this.unsubscribeFunctions.add(this.subscribePollClosed());
+    this.unsubscribeFunctions.add(this.subscribeVoteCasted());
+    this.unsubscribeFunctions.add(this.subscribeVoteChanged());
+    this.unsubscribeFunctions.add(this.subscribeVoteRemoved());
+  };
+
+  public unregisterSubscriptions = () => {
+    this.unsubscribeFunctions.forEach((cleanupFunction) => cleanupFunction());
+    this.unsubscribeFunctions.clear();
+  };
+
+  public createPoll = async (poll: CreatePollData<SCG>) => {
+    const { poll: createdPoll } = await this.client.createPoll(poll);
+
+    return new Poll({ client: this.client, poll: createdPoll });
+  };
+
+  public getPoll = async (id: string) => {
+    const { poll } = await this.client.getPoll(id);
+
+    this.setOrOverwriteInCache(poll, true);
+
+    return this.fromState(id);
```

Comment:
Why are we reaching for the request first instead of cache - I'd expect we'd be reaching for request in case of cache-miss?

Reply:
Because we've decided that `queryPolls` and `getPoll` are always going to reinitialize the state within the cache. We want them to be used independently from the rest and be able to overwrite whatever's in the cache (not by reference though, only state) when they're queried. But, perhaps we can simply always use `getPoll` and provide a fallback for whenever it's not in the cache I suppose. Wanted to keep those 2 separate though.

Reply:
There shouldn't be a need for it though, no? If the poll is in the cache then that means it's been receiving updates through event subscriptions meaning its state should closely represent the one that backend has - hydrating it at this point seems excessive.

The other way would be to return it right away and do the state hydration behind the scenes - since the state of the poll is reactive, it should not matter that it changes after it gets returned, no?

Reply:
There is a particular need for this for offline storage in RN for example. Let us consider the following scenario:

- You create a `poll` and are actively watching its progress (some votes, answers come in - w/e)
- You close the app and do something else, while in the meantime more votes and/or answers arrive
- You come back, and the first "optimistic" thing that loads is the data from the local DB (which is now out of date since those WS events never got consumed while you weren't using the app)
- If you're offline, this is no biggie since we don't expect that WS events are handled in the background
- If you're online however, the point at which a query is made for all of the channels (and their messages) you'll get all of the updated data; and it should indeed overwrite the local DB data (and store this agaiån)

Unfortunately, the `sync` API does not yet support `poll` events and so we can't particularly rely on it to do the heavy lifting for us. Whenever these potentially get added in the future, perhaps we can remove the overwriting mechanism (it's anyway hidden behind the second argument). For now, however - I believe we need it.

If it seems sketchy to you maybe we can lift the overwriting condition higher so that we can specifically trigger this only for the RN SDK (since the rest don't have offline support) ?